### PR TITLE
fix(form-core): only preserve caret if value changed

### PR DESCRIPTION
**Pull request by gerjanvangeest**
## What I did

- [x] fix: https://github.com/ing-bank/lion/issues/1415
- [x] create test suite for NativeTextFieldMixin

### CODE REVIEW

File: packages/form-core/test-suites/NativeTextFieldMixin.suite.js [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+      await el.updateComplete;
+      _inputNode.value = 'hello world';
+      _inputNode.selectionStart = 2;
+      _inputNode.selectionEnd = 2;
+      el.value = 'hey there universe';
+      expect(_inputNode.selectionStart).to.equal(2);
+      expect(_inputNode.selectionEnd).to.equal(2);
+    });
+
+    it('move focus to a next focusable element after writing some text', async () => {
```
Comment (gerjanvangeest) [inline packages/form-core/test-suites/NativeTextFieldMixin.suite.js:48]: This test also succeeds without the fix in all browsers, not yet sure how to solve it.

File: packages/form-core/src/NativeTextFieldMixin.js [hunk 2]
```diff
@@ -90,27 +96,33 @@ const NativeTextFieldMixinImplementation = superclass =>
...
+        // Only preserve caret if focused (changing selectionStart will move focus in Safari)
+        if (this.focused) {
+          // Not all elements might have selection, and even if they have the
+          // right properties, accessing them might throw an exception (like for
+          // <input type=number>)
+          try {
+            // SelectElement doesn't have selectionStart/selectionEnd
+            if (
+              !(this._inputNode instanceof HTMLSelectElement) &&
+              this._inputNode.value !== newValue
```
Comment (tlouisse) [inline packages/form-core/src/NativeTextFieldMixin.js:110]: I think the check abocve (line 100) already covered this?

File: packages/form-core/src/NativeTextFieldMixin.js [hunk 3]
```diff
@@ -90,27 +96,33 @@ const NativeTextFieldMixinImplementation = superclass =>
...
-      if (this.focused) {
-        // Not all elements might have selection, and even if they have the
-        // right properties, accessing them might throw an exception (like for
-        // <input type=number>)
-        try {
-          // SelectElement doesn't have selectionStart/selectionEnd
-          if (!(this._inputNode instanceof HTMLSelectElement)) {
-            const start = this._inputNode.selectionStart;
+      // Only set if newValue is new, fix for Safari bug: https://github.com/ing-bank/lion/issues/1415
+      if (this._inputNode.value !== newValue) {
```
Comment (tlouisse) [inline packages/form-core/src/NativeTextFieldMixin.js:100]: This could be combined into one if statement

Comment (tlouisse) [inline packages/form-core/src/NativeTextFieldMixin.js:100]: Maybe move to the top of the function to make the nexted if structure more readable:
```js
if (this._inputNode.value === newValue) { 
  return;
}
```

File: packages/form-core/test-helpers/getFormControlMembers.js [hunk 4]
```diff
@@ -12,7 +12,7 @@ export function getFormControlMembers(el) {
   // eslint-disable-next-line
   const { _inputNode, _helpTextNode, _labelNode, _feedbackNode, _allValidators } = el;
   return {
-    _inputNode,
+    _inputNode: /** @type {* & HTMLInputElement} */ (el)._inputNode,
```
Comment (tlouisse) [inline packages/form-core/test-helpers/getFormControlMembers.js:15]: I think it should be `/** @type {* & HTMLInputElement} */ (el._inputNode)`?

File: packages/textarea/test/lion-textarea-integrations.test.js [hunk 5]
```diff
@@ -1,11 +1,34 @@
...
+import { LionTextarea } from '../src/LionTextarea.js';
 
-const tagString = 'lion-textarea';
+const fieldTagString = defineCE(
+  class extends LionTextarea {
+    get slots() {
+      return {
+        ...super.slots,
+        // LionTextarea needs to have an _inputNode defined in order to work...
+        input: () => document.createElement('textarea'),
```
Comment (tlouisse) [inline packages/textarea/test/lion-textarea-integrations.test.js:16]: this is already covered by LionTextArea, or not?

File: packages/textarea/test/lion-textarea-integrations.test.js [hunk 6]
```diff
@@ -1,11 +1,34 @@
-import { runFormatMixinSuite } from '@lion/form-core/test-suites';
+import { defineCE } from '@open-wc/testing';
+import {
+  runInteractionStateMixinSuite,
+  runFormatMixinSuite,
+  runNativeTextFieldMixinSuite,
+} from '@lion/form-core/test-suites';
 
-import '@lion/textarea/define';
+import { LionTextarea } from '../src/LionTextarea.js';
```
Comment (tlouisse) [inline packages/textarea/test/lion-textarea-integrations.test.js:8]: Is there a reason to not use textarea/define anymore for this test?

File: packages/form-core/test-helpers/getFormControlMembers.js (hunk 4 above)
Comment (gerjanvangeest) [reply packages/form-core/test-helpers/getFormControlMembers.js]: Nope, the type was for the `el` element. But i changed it into `FormControlHost`

File: packages/textarea/test/lion-textarea-integrations.test.js (hunk 6 above)
Comment (gerjanvangeest) [reply packages/textarea/test/lion-textarea-integrations.test.js]: It didn't work before, but now it works again. Thanks for letting me double check it :)

File: packages/form-core/src/NativeTextFieldMixin.js (hunk 3 above)
Comment (gerjanvangeest) [reply packages/form-core/src/NativeTextFieldMixin.js]: The return made lots of tests of the `textarea` fail, i now moved the if statement one function up.

File: packages/form-core/test-suites/NativeTextFieldMixin.suite.js [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+      // TODO: this test did not fail before the fix for Safari bug: https://github.com/ing-bank/lion/issues/1415, so its not a good test.
+      const el = /** @type {NativeTextFieldClass} */ (await fixture(html`<${tag}></${tag}>`));
+      const { _inputNode } = getFormControlMembers(el);
+      await triggerFocusFor(el);
+      await el.updateComplete;
+      expect(document.activeElement).to.equal(_inputNode);
+      await sendKeys({
+        press: 'h',
+      });
+      await sendKeys({
```
Comment (tlouisse) [inline packages/form-core/test-suites/NativeTextFieldMixin.suite.js:57]: Maybe we can enable this test and instead try to check for a spy on `_setValueAndPreserveCaret` (that should not be called after the Tab)

File: packages/form-core/test-suites/NativeTextFieldMixin.suite.js (hunk 1 above)
Comment (tlouisse) [reply packages/form-core/test-suites/NativeTextFieldMixin.suite.js:51]: Maybe we can enable this test and instead try to check for a spy on _setValueAndPreserveCaret (that should not be called after the Tab)

File: packages/form-core/test-suites/NativeTextFieldMixin.suite.js [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+        press: 'Tab',
+      });
+      expect(document.activeElement).to.not.equal(_inputNode);
+      expect(setValueAndPreserveCaretSpy.calledOnce).to.be.false;
```
Comment (tlouisse) [inline packages/form-core/test-suites/NativeTextFieldMixin.suite.js:66]: Hmm, it should have callCount of 1 after the h key and after tab it should still have that callCount, right?
If calledOnce is true at the end, it seems to be called twice somehow?

Comment (gerjanvangeest) [reply packages/form-core/test-suites/NativeTextFieldMixin.suite.js:66]: Yes i thought so too, but it didn't get called...

**Review by tlouisse — APPROVED** [review]

**Pull request merged**